File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
private:
    unordered_map<char, char> map = {
        {'(', ')'},
        {'{', '}'},
        {'[', ']'}
    };
public:
    bool isValid(string s) {
        std::stack<char> stack;
        for(const auto & c : s){
            if(map.find(c) != map.end()){
                stack.push(c);
                continue;
            } else if (stack.empty()){
                return false;
            } else if (map[stack.top()] != c){
                return false;
            }
            stack.pop();
        }
        return stack.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        map: dict[int, int] = {
            ord('('):ord(')'),
            ord('{'):ord('}'),
            ord('['):ord(']'),
        }

        stack : list[str] = []

        for char in s:
            c: int = ord(char)
            if c in map:
                stack.append(c)
            elif (len(stack) == 0 or map[stack[-1]] != c):
                return False
            else : 
                stack.pop()
        
        return len(stack) == 0
                


